### PR TITLE
[FLINK-12846][table-common] Carry primary key information in TableSchema

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

The primary key is a standard meta information in SQL. And they are important information for optimization, for example, AggregateRemove, AggregateReduceGrouping and state layout optimization for TopN and Join.

So in this PR, we extend TableSchema to carry more information about primary key. 

Bridge the primary key of TableSchema to the planner metadata will be done in other PRs. 

## Brief change log

1. Add primary key to `TableSchema` with some Javadocs. 

## Verifying this change

1. Adds some tests in `TableSchemaTest` to check the primary key.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by KurtYoung** [comment]
cc @twalthr @dawidwys

**Comment by wuchong** [comment]
I dropped unique key information, because we only need to support primary key for TPC-DS optimizations.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 1]
```diff
@@ -22,32 +22,57 @@
...
 import static org.apache.flink.table.types.utils.TypeConversions.fromLegacyInfoToDataType;
+import static org.apache.flink.table.utils.TableSchemaValidation.validateNameTypeNumberEqual;
+import static org.apache.flink.table.utils.TableSchemaValidation.validateSchema;
 
 /**
- * A table schema that represents a table's structure with field names and data types.
+ * A table schema that represents a table's structure with field names, data types and
+ * constraint information (e.g. primary key, unique key).
+ *
+ * <p>Concepts about primary key and unique key:</p>
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]: Happy to help. Reading through the JavaDoc I am confused about the difference between a primary key and unique key. It sounds like the primary key is just one of the unique key constraints on a table given a special name but being labeled primary over unique does not provide any additional features or properties. Is that correct?

**Comment by wuchong** [comment]
Thanks @dawidwys for the reviewing. I agree with that the concepts of primary key on source/sink should be discussed and accepted publicly. But the background of this issue is that we would like to support more optimizations for TPC-DS where the sources have primary key information in Hive metastore. So we would like to only expose the primary key interface to `CatalogTable`, to let the Hive catalog pass the primary key information which is defined by Hive DDL to Flink. If a primary key is defined on Flink DDL or `TableSource#getTableSchema` or `TableSink#getTableSchema`, an unsupported exception will be thrown. In this way, I think the primary key concepts is simple and clear which is similar to [Hive's definition](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL).

But I'm not sure whether this still requires a FLIP. What do you think?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 2]
```diff
@@ -246,7 +287,28 @@ public DataType toRowDataType() {
...
+		return new ArrayList<>(watermarkSpecs);
+	}
+
+	/**
+	 * Returns primary key defined on the table. A primary key is consist of single or
+	 * multiple field names. The returned list will be empty if no primary key is defined.
+	 *
+	 * <p>See the {@link TableSchema} class javadoc for more definition about primary key.
+	 */
+	public List<String> getPrimaryKey() {
```
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]: I think having a dedicated `PrimaryKey` class is a good idea.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaValidation.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+				column.getType(),
+				"");
+		}
+
+		// primary key and unique key requires the key field is top-level (non-nested)
+		Map<String, TableColumn> nameToColumn = columns.stream()
+			.collect(Collectors.toMap(TableColumn::getName, Function.identity()));
+
+		// validate primary key
+		for (String key : primaryKey) {
```
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/TableSchemaValidation.java]: Yes. I think we should validate the key fields are unique.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 4]
```diff
@@ -22,32 +22,49 @@
...
+ * A table schema that represents a table's structure with field names, data types, watermark
+ * information and constraint information (e.g. primary key).
+ *
+ * <p>Concepts about primary key:</p>
+ * <ul>
+ *     <li>
+ *         Primary key can be consist of single or multiple columns (fields).
+ *     </li>
+ *     <li>
+ *         A primary key on source will be simply trusted, we won't validate the constraint.
```
Comment (sjwiesman) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]:
```suggestion
 *         A primary key on a source table will simply be trusted, we won't validate the constraint.
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 5]
```diff
@@ -22,32 +22,49 @@
...
+ *         The primary key information will then be used for query optimization. If a bounded
+ *         or unbounded table source defines any primary key, it must contain a unique value
+ *         for each row of data. You cannot have two records having the same value of that field(s).
```
Comment (sjwiesman) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]:
```suggestion
 *         for each row of data. If your table contains multiple rows with the same primary key values
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 6]
```diff
@@ -22,32 +22,49 @@
...
+ *         Otherwise, the result of query might be wrong.
```
Comment (sjwiesman) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]:
```suggestion
 *         then the result of the query might be wrong.
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 7]
```diff
@@ -22,32 +22,49 @@
...
+ *     </li>
+ *     <li>
+ *         A primary key on sink is a weak constraint. Currently, we won't validate the constraint,
```
Comment (sjwiesman) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]:
```suggestion
 *         A primary key on sink is a weak constraint and is not validated by Flink. However, 
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 8]
```diff
@@ -22,32 +22,49 @@
...
+ *         but we may add some check in the future to validate whether the primary key of the query
```
Comment (sjwiesman) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java]:
```suggestion
 *         we may add some checks in the future to validate whether the primary key of the query
```

**Comment by dawidwys** [comment]
Hi @wuchong.
First of all thank you for outlining the original purpose of this change. I am a bit hesitant though to introduce a new ground concept to a core part of the Table ecosystem without a good understanding of it. That's why I would appreciate an agreement within community. I do want to emphasize I definitely want to help with finding a solution that we can introduce in 1.10 that can improve the TPC-DS performance.

I did a bit of research on the `PRIMARY & UNIQUE` keys. I have one concern regarding the semantics of a `PRIMARY KEY`. What I found out is that according to SQL standard & most of RDBM systems, `PRIMARY KEY` constraint enforces `NOT NULL` constraint on all columns that are part of the `PRIMARY KEY`. That's not the case for `UNIQUE` keys, but there it is assumed that null values are not equal regarding that constraint (you may have multiple rows having a null unique key). Unfortunately this is not reflected in Hive. In hive (at least pre 3.0) there was/is no support of `NOT NULL` types. Therefore `PRIMARY KEY` in hive is actually not a `PRIMARY KEY` in a SQL standard sense. I think just for that issue it would be good to bring it to the ML.

**Pull request closed without merging**